libtest: Print the total time taken to execute a test suite

jakoschiko opened the pull request:
Print the total time taken to execute a test suite by default, without any kind of flag.

Closes #75660

# Example
```
anon@anon:~/code/rust/example$ cargo test
   Compiling example v0.1.0 (/home/anon/code/rust/example)
    Finished test [unoptimized + debuginfo] target(s) in 0.18s
     Running target/debug/deps/example-745b64d3885c3565

running 3 tests
test tests::foo ... ok
test tests::bar ... ok
test tests::baz ... ok

test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; 1.2s

   Doc-tests example

running 3 tests
test src/lib.rs - foo (line 3) ... ok
test src/lib.rs - bar (line 11) ... ok
test src/lib.rs - baz (line 19) ... ok

test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; 1.3s
```

```
anon@anon:~/code/rust/example$ cargo test -- --format terse
    Finished test [unoptimized + debuginfo] target(s) in 0.08s
     Running target/debug/deps/example-745b64d3885c3565

running 3 tests
...
test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; 1.2s

   Doc-tests example

running 3 tests
...
test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; 1.3s
```

```
anon@anon:~/code/rust/example$ cargo test -- --format json -Z unstable-options
   Compiling example v0.1.0 (/home/anon/code/rust/example)
    Finished test [unoptimized + debuginfo] target(s) in 0.25s
     Running target/debug/deps/example-745b64d3885c3565
{ "type": "suite", "event": "started", "test_count": 3 }
{ "type": "test", "event": "started", "name": "tests::bar" }
{ "type": "test", "event": "started", "name": "tests::baz" }
{ "type": "test", "event": "started", "name": "tests::foo" }
{ "type": "test", "name": "tests::foo", "event": "ok" }
{ "type": "test", "name": "tests::bar", "event": "ok" }
{ "type": "test", "name": "tests::baz", "event": "ok" }
{ "type": "suite", "event": "ok", "passed": 3, "failed": 0, "allowed_fail": 0, "ignored": 0, "measured": 0, "filtered_out": 0, "exec_time": "1.2s" }
   Doc-tests example
{ "type": "suite", "event": "started", "test_count": 3 }
{ "type": "test", "event": "started", "name": "src/lib.rs - bar (line 11)" }
{ "type": "test", "event": "started", "name": "src/lib.rs - baz (line 19)" }
{ "type": "test", "event": "started", "name": "src/lib.rs - foo (line 3)" }
{ "type": "test", "name": "src/lib.rs - foo (line 3)", "event": "ok" }
{ "type": "test", "name": "src/lib.rs - bar (line 11)", "event": "ok" }
{ "type": "test", "name": "src/lib.rs - baz (line 19)", "event": "ok" }
{ "type": "suite", "event": "ok", "passed": 3, "failed": 0, "allowed_fail": 0, "ignored": 0, "measured": 0, "filtered_out": 0, "exec_time": "1.3s" }
```

rust-highfive commented:
r? @LukasKalbertodt

(rust_highfive has picked a reviewer for you, use r? to override)

jakoschiko commented:
Setting the time to zero sounds reasonable.

The [documentation](https://rustc-dev-guide.rust-lang.org/tests/adding.html) says rustc is run with `-Z ui-testing` when compiling the UI tests. Can I check that flag at compile time? This would be better than adding another `-Z` flag to libtest.

jakoschiko commented:
> Can I check that flag at compile time?

I found no clues in the source code that this is possible. Therefore I implemented `-Z ui-testing` for libtest.
- Does this need further documentation?
- Is it confusing that it uses the same name as rustc's `-Z ui-testing`?

LukasKalbertodt commented:
Great idea; that sounds very useful!

I think a better way to fix the "test output" problem is to use custom normalization of the output, as described [here](https://rustc-dev-guide.rust-lang.org/tests/adding.html#normalization). I haven't looked into it in detail, but it seems like we can regex-replace the actual output before it is compared to the `.stdout` file. (If you need help with this, I can of course look into it more thoroughly.)
I would much prefer that over an unstable flag, especially since the flag is responsible for a large part of code changes in this PR.

And I think the output should say something like `finished in 3s` or something like that, not just print the time.

jakoschiko commented:
Normalization works great! I also changed the output as proposed, this helped to find the text that needs to be normalized.
```
anon@anon:~/code/rust/example$ cargo +stage1 test
   Compiling example v0.1.0 (/home/anon/code/rust/example)
    Finished test [unoptimized + debuginfo] target(s) in 0.26s
     Running target/debug/deps/example-745b64d3885c3565

running 3 tests
test tests::foo ... ok
test tests::bar ... ok
test tests::baz ... ok

test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 1.20s

   Doc-tests example

running 3 tests
test src/lib.rs - foo (line 3) ... ok
test src/lib.rs - bar (line 11) ... ok
test src/lib.rs - baz (line 19) ... ok

test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 1.32s
```

jakoschiko commented:
Yet another failing test:
`src/test/run-make-fulldeps/libtest-json` runs libtest with JSON format via a Makefile, redirects the output into a file, checks for valid JSON and compares it with a file containing the expected output. Problem: The actual output is nondeterministic.

Possible solutions:
- Don't add the execution time to the JSON output. This was not requested, but I thought it makes sense.
- Reintroduce an unstable flag that ensures deterministic output. This was already done and required too much changes.
- Use a CLI tool to normalize the actual output. Which tools are allowed to be used?

Any advice?

LukasKalbertodt commented:
Looks much better now!

The second test failure is unfortunate. I am not sure how to best fix this :/ 
I would tend to believe that the execution time in the JSON output isn't as important, but I am not sure if there are some guidelines on what to include there. I also have no CLI idea what tools can be used in that test.
While I don't particularly like the unstable flag, it might still be a valid solution.

But in general, I'm not sure I am the correct person to review this (doesn't seem like a libs issue). Maaaybe @rust-lang/devtools can help out?

Mark-Simulacrum commented:
I suspect that normalizing the JSON output shouldn't be too hard -- we can use anything on CI, probably, but something liked `sed` should do the trick and I think is likely already used in other run-make tests.

jakoschiko commented:
Thanks, I fixed the test with `sed`.

And yes, `sed` is already used, for example in `src/test/run-make-fulldeps/stable-symbol-names/Makefile`.

Manishearth commented:
I would say libtest is a libs team concern

LukasKalbertodt commented:
> I would say libtest is a libs team concern

Well then!

@bors r+

Thanks again for the PR!

bors commented:
:pushpin: Commit f8d0857135c216621c39edefbb2ecddb26d2208a has been approved by `LukasKalbertodt`

<!-- @bors r=LukasKalbertodt f8d0857135c216621c39edefbb2ecddb26d2208a -->
<!-- homu: {"type":"Approved","sha":"f8d0857135c216621c39edefbb2ecddb26d2208a","approver":"LukasKalbertodt"} -->

matklad commented:
This is not gated, right? Then this should hit relnotes

bors commented:
:hourglass: Testing commit f8d0857135c216621c39edefbb2ecddb26d2208a with merge 9d4918d9118417d0a46db7131fc474ec0ac3ee9b...
<!-- homu: {"type":"BuildStarted","head_sha":"f8d0857135c216621c39edefbb2ecddb26d2208a","merge_sha":"9d4918d9118417d0a46db7131fc474ec0ac3ee9b"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1047494468)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1047494468","builder_name":"checks-actions"} -->

jakoschiko commented:
Some tests failed on `test-various, ubuntu-latest-xl`. It looks like `std::time::Instant::now` (used by this PR for measuring time) is not implemented for wasm. Example output from one of the tests:
```
---- [ui] ui/issues/issue-16597-empty.rs stdout ----

error: test run failed!
status: exit code: 101
command: "/node-v14.4.0-linux-x64/bin/node" "/checkout/src/etc/wasm32-shim.js" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/issues/issue-16597-empty/a.wasm"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
RuntimeError: unreachable
    at __rust_start_panic (wasm-function[440]:0x367e5)
    at rust_panic (wasm-function[429]:0x364b0)
    at _ZN3std9panicking20rust_panic_with_hook17hd145b35fc2e3a3cfE (wasm-function[424]:0x36193)
    at _ZN3std9panicking11begin_panic28_$u7b$$u7b$closure$u7d$$u7d$17hb1773171f7f93b68E (wasm-function[412]:0x35e10)
    at _ZN3std10sys_common9backtrace26__rust_end_short_backtrace17hf7e9c7af0f61d223E (wasm-function[411]:0x35dd7)
    at _ZN3std9panicking11begin_panic17hffdce10487e9f2d5E (wasm-function[306]:0x3125a)
    at _ZN3std3sys4wasm4time7Instant3now17h0cd4438a2c9238d7E (wasm-function[404]:0x35bc2)
    at _ZN3std4time7Instant3now17hfc7bdb933c65609fE (wasm-function[398]:0x359b0)
    at _ZN4test7console17run_tests_console17h9236dc85784dfce6E (wasm-function[134]:0x10fd5)
    at _ZN4test9test_main17hf7027953a2d207bdE (wasm-function[182]:0x2198f)
    at _ZN4test16test_main_static17hd410e3e9c2ad1d88E (wasm-function[183]:0x22a9a)
    at _ZN17issue_16597_empty4main17hce00e52b5711c0baE (wasm-function[1]:0x545)
    at _ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17h527334a16a9115efE (wasm-function[0]:0x521)
    at _ZN3std2rt10lang_start28_$u7b$$u7b$closure$u7d$$u7d$17hd43ebc70162051f2E.llvm.11756281066959917856 (wasm-function[3]:0x5ab)
    at _ZN3std2rt19lang_start_internal17h1326eb4f191ee6aaE (wasm-function[430]:0x3657a)
    at main (wasm-function[2]:0x57b)
    at Object.<anonymous> (/checkout/src/etc/wasm32-shim.js:20:20)
    at Module._compile (internal/modules/cjs/loader.js:1200:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1220:10)
    at Module.load (internal/modules/cjs/loader.js:1049:32)

------------------------------------------
```
Any advice?

LukasKalbertodt commented:
One could of course [ignore those tests](https://rustc-dev-guide.rust-lang.org/tests/adding.html#ignoring-tests) on wasm targets, but it looks like these test failures should not be ignored, but are rather important.

So we can run the test harness on wasm? Is that the status quo? If so, we can't just break it, of course.
But I am confused as the harness already uses quite a few  `Instant::now` that doesn't seem to be problematic? There are two `cfg(target_arch =  "wasm32")` in the code, but both seem unrelated to the uses of `Instant`. 

One possibility would be to just not measure the time on wasm targets. But I am not sure if that's really the best way forward :/

crlf0710 commented:
@jakoschiko Ping from triage, any updates on this?

jakoschiko commented:
@LukasKalbertodt @crlf0710

I'm still alive. These issues are blocking me:

1. I don't know how to run the wasm test suite locally. So I couldn't debug the problem.
2. I don't know what's the best way forward. These are the possibilities I know:
  a. Ignore the tests. I assume that's not an option.
  b. Implement `Instant` for wasm. I assume that's not within the scope of this PR. I don't even understand the whole situation. Why is stdlib partially implemented for wasm?
  c. Don't print the time when running the test suite on wasm. That sounds preferable. But I assume that there are or will be UI tests that will fail only on wasm. And tweaking these tests only for wasm sounds annoying.
  d. Provide a custom measurement utility for libtest that always returns 0 on wasm. And then print `0.00s`. This sounds like a hacky solution, but it would work.
3. What's with the other usages of `Instant` in libtest? Should they be fixed, too? I couldn't debug them, but I assume they will panic.
   a. `bench` is using `Instant`, but I think it's ok because it's unstable.
   b. `report-time`/`ensure-time` is using `Instant`, but I think it's ok because it's unstable.
   c. The timeout feature is using `Instant`, but I think it's ok because the timeout feature is only available for asynchronous test runs and there's a `cfg!(target_arch = "wasm32")` that forces the test to be run synchronously on wasm.

If someone would agree that it's a good idea to implement 2.d. and ignore 3., then I will do it. Or are there better options?

Mark-Simulacrum commented:
I don't think 0 makes sense on wasm, we should probably just not print (or say something like `<not supported>`).

If you're able to use docker, then you can run `src/ci/docker/run.sh test-various` to get an identical image to the CI builder that failed. I think you should be able to do `x.py check --target wasm32-unknown-unknown library/test` to make sure your changes at least build.

ehuss commented:
Docker is a good idea if you want exactly the same as CI and don't want to mess with trying to run directly.  If you want to try without docker, you can install node.js (v14), set `lld = true` in config.toml, and run `./x.py test --target wasm32-unknown-unknown src/test/run-make src/test/ui src/test/compile-fail src/test/mir-opt src/test/codegen-units library/core`.

I would also vote to just not show the time.

jakoschiko commented:
Ok, I can run the tests with docker now. 

I want to add a UI test that runs only on wasm and proves that the time will not be printed. The available `src/test/ui/test-*` tests are disabled for wasm because wasm has no panic support. So I want to add a test that doesn't panic.

There are some other tests that uses `// only-wasm32`, e.g. `ui/wasm-custom-section-relocations.rs`, but these tests will still be ignored:
```
Check compiletest suite=ui mode=ui (x86_64-unknown-linux-gnu -> wasm32-unknown-unknown)
...
test [ui] ui/wasm-custom-section-relocations.rs ... ignored
```

Any ideas?

jakoschiko commented:
A summary of the current state:
- The execution time is measured and printed only on non-wasm targets.
- I fixed a lot of UI tests by normalizing their outputs.
- I added a UI test for testing the test harness output with only passed tests. It won't run on wasm targets because of the wasm-check-run-results-exception I mentioned above.
- I didn't add any tests for the test harness output on wasm targets. Even without the wasm-check-run-results-exception, the actual output produced by compiletest is always empty. I assume that check-run-results-tests on wasm are simply not supported by compiletest.
- The wasm tests on test-various are fixed. There are still some failing on x86_64-unknown-linux-musl. These tests lie in `net::tcp::tests` and fail because an address is not available. I hope this is unrelated to my changes.

Waiting for another review.

jakoschiko commented:
Bump @LukasKalbertodt @crlf0710
Is there anything I can do?

Dylan-DPC commented:
r? @Amanieu

m-ou-se commented on library/test/src/formatters/json.rs:
```diff
@@ -156,24 +156,31 @@ impl<T: Write> OutputFormatter for JsonFormatter<T> {
         ))?;
 
+        if let Some(ref exec_time) = state.exec_time {
+            let time_str = format!(", \"exec_time\": \"{}\"", exec_time);
```
```suggestion
            let time_str = format!(", \"exec_time\": {}", exec_time.0.as_secs_f64());
```

m-ou-se reviewed (commented):
Hey, thanks for working on this! This all looks great.

One minor suggestion: Since JSON supports numbers, I think it'd be better to do `"exec_time": 1.2` instead of `"exec_time": "1.2s"`. Makes it a bit easier to consume for other programs.

jakoschiko commented:
> One minor suggestion: Since JSON supports numbers, I think it'd be better to do `"exec_time": 1.2` instead of `"exec_time": "1.2s"`. Makes it a bit easier to consume for other programs.

Sounds reasonable, but:
- Is the unit of time obvious? Should I also change the key to `"exec_time_s"`?
- I used the same format for the `report-time` flag (#64663). Is it okay that I use a different format this time? Should I make the same change for `report-time`?

m-ou-se commented:
I think the unit is obvious. Seconds is the iso unit for time. Especially as a float, it's very uncommon to use anything other than seconds.

It'd be good if it's consistent with `report-time`, but since that one is still unstable, it'd be fine to change that (now or later).

jakoschiko commented:
I changed the JSON output as suggested: The key "exec_time" has now a number value. I made the change for the execution time of unit tests (available via `report-time` flag) and the test suite (available by default).

jakoschiko commented:
I rebased the branch onto the current master and fixed some new ui tests whose output needed to be normalized.

m-ou-se commented:
Thank you!

@bors r+

bors commented:
:pushpin: Commit 46993efcfb2cd08e83469d610b00449bae01b7d9 has been approved by `m-ou-se`

<!-- @bors r=m-ou-se 46993efcfb2cd08e83469d610b00449bae01b7d9 -->
<!-- homu: {"type":"Approved","sha":"46993efcfb2cd08e83469d610b00449bae01b7d9","approver":"m-ou-se"} -->

m-ou-se commented:
This failed in https://github.com/rust-lang/rust/pull/79317: https://github.com/rust-lang-ci/rust/runs/1439245145

@bors r-

jakoschiko commented:
Interesting, there are 3 tests in the cargo submodule that also check the libtest output.

I don't know how to fix or even run them yet. I'll investigate it.

Given that cargo has its own repo, in which order this need to be merged?

Mark-Simulacrum commented:
You'll likely want to change the cargo repository to accept both past and present outputs, and file a PR against rust-lang/cargo. Once that's merged I would coordinate with @ehuss to get the submodule bump done in rust-lang/rust, and then this can be merged once the submodule is bumped.

jakoschiko commented:
Thanks! I will do exactly that.

jakoschiko commented:
@ehuss 

I need a bump of rust-lang/cargo in rust-lang/rust. I'm not familiar with this. Who is when bumping the submodules? Can I do something besides asking?

ehuss commented:
@jakoschiko I've been away for a bit, so I'm trying to catch up on the backlog of notifications.  I'll try to do an update very soon (today or tomorrow).

jakoschiko commented:
Ok, maybe there is more to do. I searched the rust repo for more libtest output.

1) The following documentations contains libtest output:
     - `src/doc/book`
     - `src/doc/unstable-book`
     - `src/doc/rust-by-example`
     - `src/tools/cargo/src/doc`

    Do we want to update them?

2) And there are 5 tests in `src/tools/miri/test-cargo-miri` with `test.stdout.ref1` to `test.stdout.ref5`. I suppose miri doesn't even support `std::time::Instant` (similar to wasm32)? Then I should disable the execution time for miri, too.

3) And there is `src/tools/rust-analyzer/crates/ide/src/completion/generated_features.rs`. It's a generated file, but I don't know if I should be worried.

ehuss commented:
I wouldn't worry about the documentation.  It looks like the rust-analyzer thing is just some docs generated from the unstable-book, so I wouldn't worry about that, either.

I'm not sure about miri.  It looks like the tests run on rust-lang/rust should be OK.  If the `test-cargo-miri` is broken, it will need to be updated after the PR is merged.  It looks like miri has some shims for time, though I'm not familiar with it.

jakoschiko commented:
The miri test is indeed failing:
```
~/code/rust/miri/test-cargo-miri$ ./run-test.py 
## Running `cargo miri` tests
A libstd for Miri is now available in `/home/anon/.cache/miri/HOST`.
Testing `cargo miri run` (no isolation)...
Testing `cargo miri run` (with arguments and target)...
Testing `cargo miri run` (subcrate, no ioslation)...
Testing `cargo miri test`...
--- BEGIN stdout ---
--- END stdout ---
--- BEGIN stderr ---
error: unsupported operation: `clock_gettime` not available when isolation is enabled
   --> /home/anon/code/rust/rust/library/std/src/sys/unix/time.rs:371:22
    |
371 |         cvt(unsafe { libc::clock_gettime(clock, &mut t.t) }).unwrap();
    |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `clock_gettime` not available when isolation is enabled
    |
    = help: pass the flag `-Zmiri-disable-isolation` to disable isolation
...
```

I think there are these two options:
- Disable the execution time output via conditional compilation. I already did this for wasm32, but I'm not sure how to do this for miri, `target_arch` doesn't work. This must be done before merging this PR.
- Use the flag `-Zmiri-disable-isolation` when running the miri test and normalize the output. If this can be done, it can be done later, but it's over my head.

m-ou-se commented:
@RalfJung see last comment. ^ Is it a problem for Miri if libtest starts using Instant::now()/elapsed() to report the time tests take? If so, does that need a workaround in this PR, or can/should that be worked around separately?

RalfJung commented:
I'd prefer if we could keep `cargo miri test` working with isolation.

@jakoschiko  I recommend you use `cfg(miri)` to disable the measurement when building for Miri, is that possible? You mentioned youa re already using conditional compilation for wasm32 so I hope this further condition can be easily added.

bors commented:
:umbrella: The latest upstream changes (presumably #79441) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

jakoschiko commented:
@RalfJung Yeah, it seems to work, thanks!

@m-ou-se I rebased onto the current master, squashed the commits and added `cfg(miri)` to `console.rs`. IMHO it's ready to be merged.

m-ou-se commented:
Thanks a lot, @jakoschiko!

@bors r+

bors commented:
:pushpin: Commit becc4c243d60feed8cd5d4096b676d6c6c4a7797 has been approved by `m-ou-se`

<!-- @bors r=m-ou-se becc4c243d60feed8cd5d4096b676d6c6c4a7797 -->
<!-- homu: {"type":"Approved","sha":"becc4c243d60feed8cd5d4096b676d6c6c4a7797","approver":"m-ou-se"} -->

bors commented:
:hourglass: Testing commit becc4c243d60feed8cd5d4096b676d6c6c4a7797 with merge 50cc57bf9e105b40f4087de04bafe4a617831ebb...
<!-- homu: {"type":"BuildStarted","head_sha":"becc4c243d60feed8cd5d4096b676d6c6c4a7797","merge_sha":"50cc57bf9e105b40f4087de04bafe4a617831ebb"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1463436829)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1463436829","builder_name":"checks-actions"} -->

m-ou-se commented:
The `ui/test-passed.rs` test failed on wasm:

```diff
--- [ui] ui/test-passed.rs stdout ----
diff of run.stdout:

3	test it_works ... ok
4	test it_works_too ... ok
5	
-	test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in $TIME
+	test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
7	
8	
```

There's probably more tests that would fail in the same way on wasm.

You could either add `// ignore-emscripten` to those tests, or remove all of `; finished in ..` in the output normalization, depending on what the test is testing.

jakoschiko commented:
The test `ui/test-passed.rs` failed on wasm32, ubuntu-latest-xl. It passed the last time.

I wrote earlier:

>     * I added a UI test for testing the test harness output with only passed tests. It won't run on wasm targets because of the wasm-check-run-results-exception I mentioned above.
> 
>     * I didn't add any tests for the test harness output on wasm targets. Even without the wasm-check-run-results-exception, the actual output produced by compiletest is always empty. I assume that check-run-results-tests on wasm are simply not supported by compiletest.

Maybe this changed in the meantime?

jakoschiko commented:
- I rebased onto the current master
- I added `// ignore-wasm32` to the test `ui/test-passed.rs`, similar to the `!cfg!(target_arch = "wasm32")` in `console.rs`
- I added the test `ui/test-passed-wasm.rs` with `// only-wasm32` for testing the output without execution time
- I used the wasm32 docker image to check that the tests actually work

@m-ou-se Let's try it again?

m-ou-se commented:
Thanks!

@bors r+

bors commented:
:pushpin: Commit 470c059e691a4e7befa5e2a7ad0110a2b0055b50 has been approved by `m-ou-se`

<!-- @bors r=m-ou-se 470c059e691a4e7befa5e2a7ad0110a2b0055b50 -->
<!-- homu: {"type":"Approved","sha":"470c059e691a4e7befa5e2a7ad0110a2b0055b50","approver":"m-ou-se"} -->

bors commented:
:hourglass: Testing commit 470c059e691a4e7befa5e2a7ad0110a2b0055b50 with merge 9564843645246ccfcfcefc5e294bb964bcecc1da...
<!-- homu: {"type":"BuildStarted","head_sha":"470c059e691a4e7befa5e2a7ad0110a2b0055b50","merge_sha":"9564843645246ccfcfcefc5e294bb964bcecc1da"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1466743174)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1466743174","builder_name":"checks-actions"} -->

m-ou-se commented:
Almost there! All tests passed on all platforms, except on one Apple runner where `sed` doesn't have the `-r` flag.

m-ou-se commented:
@bors r+

bors commented:
:pushpin: Commit fe3daea10414fd9ba1e956ee1bcf23db07c8a7c2 has been approved by `m-ou-se`

<!-- @bors r=m-ou-se fe3daea10414fd9ba1e956ee1bcf23db07c8a7c2 -->
<!-- homu: {"type":"Approved","sha":"fe3daea10414fd9ba1e956ee1bcf23db07c8a7c2","approver":"m-ou-se"} -->

jakoschiko commented:
I hope this will work:
```
$ echo -e '{ "foo": "foo", "bar": 42, "exec_time": 0 }
> { "foo": "foo", "bar": 42, "exec_time": 10 }
> { "foo": "foo", "bar": 42, "exec_time": 20 }
> { "foo": "foo", "bar": 42, "exec_time": 0.1 }
> { "foo": "foo", "bar": 42, "exec_time": 0.12 }
> { "foo": "foo", "bar": 42, "exec_time": 10.12 }
> { "foo": "foo", "bar": 42, "exec_time": 200.12 }' |\
> sed 's/\"exec_time\": [0-9]\+\(\.[0-9]\+\)\?/\"exec_time\": \$$TIME/'
{ "foo": "foo", "bar": 42, "exec_time": $$TIME }
{ "foo": "foo", "bar": 42, "exec_time": $$TIME }
{ "foo": "foo", "bar": 42, "exec_time": $$TIME }
{ "foo": "foo", "bar": 42, "exec_time": $$TIME }
{ "foo": "foo", "bar": 42, "exec_time": $$TIME }
{ "foo": "foo", "bar": 42, "exec_time": $$TIME }
{ "foo": "foo", "bar": 42, "exec_time": $$TIME }
```

bors commented:
:hourglass: Testing commit fe3daea10414fd9ba1e956ee1bcf23db07c8a7c2 with merge 7ea7f3f48fdb4afd756d4ee77a4fbb15d494e28e...
<!-- homu: {"type":"BuildStarted","head_sha":"fe3daea10414fd9ba1e956ee1bcf23db07c8a7c2","merge_sha":"7ea7f3f48fdb4afd756d4ee77a4fbb15d494e28e"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1466988150)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1466988150","builder_name":"checks-actions"} -->

m-ou-se commented:
Spurious error:

```
error: failed to download from `https://crates.io/api/v1/crates/rustc-demangle/0.1.18/download`

Caused by:
  [55] Failed sending data to the peer (Connection died, tried 5 times before giving up)
```

@bors retry

bors commented:
:hourglass: Testing commit fe3daea10414fd9ba1e956ee1bcf23db07c8a7c2 with merge ed9f6cafa75fbe0e18198e6fabcdf75e7d5e1b1e...
<!-- homu: {"type":"BuildStarted","head_sha":"fe3daea10414fd9ba1e956ee1bcf23db07c8a7c2","merge_sha":"ed9f6cafa75fbe0e18198e6fabcdf75e7d5e1b1e"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1467494948)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1467494948","builder_name":"checks-actions"} -->

m-ou-se commented:
Oh, apparently POSIX doesn't even require `\+` to be supported.

I think this would work:
```sh
sed 's/"exec_time": [0-9.]*/"exec_time": $$TIME/'
```

jakoschiko commented:
@m-ou-se Thanks, I changed the latest commit accordingly.

Dylan-DPC commented:
@bors r=@m-ou-se

bors commented:
:pushpin: Commit 322f53d8a8da14da5371ca040ade5497112178cf has been approved by `m-ou-se`

<!-- @bors r=m-ou-se 322f53d8a8da14da5371ca040ade5497112178cf -->
<!-- homu: {"type":"Approved","sha":"322f53d8a8da14da5371ca040ade5497112178cf","approver":"m-ou-se"} -->

bors commented:
:hourglass: Testing commit 322f53d8a8da14da5371ca040ade5497112178cf with merge 6add378d6b77a727697f2920a9cc85a63766297e...
<!-- homu: {"type":"BuildStarted","head_sha":"322f53d8a8da14da5371ca040ade5497112178cf","merge_sha":"6add378d6b77a727697f2920a9cc85a63766297e"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1469278018)
Approved by: m-ou-se
Pushing 6add378d6b77a727697f2920a9cc85a63766297e to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"m-ou-se","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1469278018"},"merge_sha":"6add378d6b77a727697f2920a9cc85a63766297e"} -->

jakoschiko commented:
Thanks for the help!

m-ou-se commented:
@jakoschiko Thanks for all your work to make this happen!
